File: frontend/src/types/document.ts
export type DocumentStatus =
  | 'PENDING'
  | 'PROCESSING'
  | 'PROCESSED'
  | 'COMPLETED'
  | 'UPLOADED'
  | 'FAILED';

export interface DocumentPage {
  id: string;
  page_number: number;
  raw_text: string;
  layout_blocks?: any[];
}

export interface Document {
  id: string;
  filename: string;
  file_path: string;
  file_hash: string;
  file_size: number;
  num_pages: number;
  page_count?: number;
  status: DocumentStatus;
  error_message?: string;
  metadata_json?: Record<string, any>;
  created_at: string;
  updated_at: string;
  fact_count?: number;
  pages?: DocumentPage[];
}

File: backend/app/models/document.py
"""Document and DocumentPage database models."""
import enum
from datetime import datetime
from sqlalchemy import Column, String, Integer, DateTime, Enum, Text, ForeignKey, JSON
from sqlalchemy.orm import relationship
from app.db.base import Base


class DocumentStatus(str, enum.Enum):
    PENDING = "PENDING"
    PROCESSING = "PROCESSING"
    PROCESSED = "PROCESSED"
    FAILED = "FAILED"


class Document(Base):
    __tablename__ = "documents"

    id = Column(String(36), primary_key=True, index=True)
    filename = Column(String(255), nullable=False)
    file_path = Column(String(512), nullable=False)
    file_hash = Column(String(64), nullable=False, unique=True, index=True)
    file_size = Column(Integer, nullable=False)
    num_pages = Column(Integer, default=0)
    status = Column(Enum(DocumentStatus), default=DocumentStatus.PENDING, index=True)
    error_message = Column(Text, nullable=True)
    metadata_json = Column(JSON, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    pages = relationship("DocumentPage", back_populates="document", cascade="all, delete-orphan")
    facts = relationship("Fact", back_populates="document", cascade="all, delete-orphan")
    extraction_logs = relationship("ExtractionFailure", back_populates="document", cascade="all, delete-orphan")


class DocumentPage(Base):
    __tablename__ = "document_pages"

    id = Column(String(36), primary_key=True, index=True)
    document_id = Column(String(36), ForeignKey("documents.id", ondelete="CASCADE"), nullable=False, index=True)
    page_number = Column(Integer, nullable=False)  # 1-indexed
    raw_text = Column(Text, nullable=False)
    layout_blocks = Column(JSON, nullable=True)  # Bounding boxes, lines, tables
    page_hash = Column(String(64), nullable=True)

    document = relationship("Document", back_populates="pages")
    facts = relationship("Fact", back_populates="page")
